File: ts/src/presets.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Copyright (c) 2026 Yey Boats Project. See LICENSE and COMMERCIAL.md.

import type { Node } from "./types";

export type PresetExpander = (slots: string[]) => Node;

function req(slots: string[], n: number): string[] {
  if (slots.length !== n) throw new Error(`preset expects ${n} slots, got ${slots.length}`);
  return slots;
}

// Registry of named presets. Each expands to a Node (which may itself
// contain further presets — expand() recurses). Add new presets here.
export const PRESETS: Record<string, PresetExpander> = {
  full: (s) => ({ element: req(s, 1)[0] }),
  "hero-split": (s) => {
    const [a, b, c] = req(s, 3);
    return { dir: "row", children: [{ element: a }, { dir: "col", children: [{ element: b }, { element: c }] }] };
  },
};

export function expand(n: Node): Node {
  if ("preset" in n) {
    const fn = PRESETS[n.preset];
    if (!fn) throw new Error(`unknown preset: ${n.preset}`);
    return expand(fn(n.slots ?? []));
  }
  if ("children" in n) return { ...n, children: n.children.map(expand) };
  if ("cells" in n) return { ...n, cells: n.cells.map(expand) };
  return n;
}

export function countTiles(n: Node): number {
  if ("element" in n) return 1;
  if ("children" in n) return n.children.reduce((a, c) => a + countTiles(c), 0);
  if ("cells" in n) return n.cells.reduce((a, c) => a + countTiles(c), 0);
  return 0; // a preset node (should not appear after expand)
}

export function depth(n: Node): number {
  if ("element" in n) return 1;
  const kids = "children" in n ? n.children : "cells" in n ? n.cells : [];
  return kids.length ? 1 + Math.max(...kids.map(depth)) : 1;
}

File: py/src/yey_boats_midl/presets.py
# SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
# Copyright (c) 2026 Yey Boats Project. See LICENSE and COMMERCIAL.md.
"""Layout presets: expand / countTiles / depth — mirrors ts/src/presets.ts."""
from __future__ import annotations

from typing import Any, Callable, Dict, List

Node = Dict[str, Any]


def _req(slots: List[str], n: int) -> List[str]:
    if len(slots) != n:
        raise ValueError(f"preset expects {n} slots, got {len(slots)}")
    return slots


def _full(s: List[str]) -> Node:
    return {"element": _req(s, 1)[0]}


def _hero_split(s: List[str]) -> Node:
    a, b, c = _req(s, 3)
    return {
        "flow": "row",
        "children": [
            {"element": a},
            {"flow": "col", "children": [{"element": b}, {"element": c}]},
        ],
    }


PRESETS: Dict[str, Callable[[List[str]], Node]] = {
    "full": _full,
    "hero-split": _hero_split,
}


def expand(n: Node) -> Node:
    if "preset" in n:
        fn = PRESETS.get(n["preset"])
        if fn is None:
            raise ValueError(f"unknown preset: {n['preset']}")
        return expand(fn(n.get("slots") or []))
    if "children" in n:
        return {**n, "children": [expand(c) for c in n["children"]]}
    if "cells" in n:
        return {**n, "cells": [expand(c) for c in n["cells"]]}
    return n


def count_tiles(n: Node) -> int:
    if "element" in n:
        return 1
    if "children" in n:
        return sum(count_tiles(c) for c in n["children"])
    if "cells" in n:
        return sum(count_tiles(c) for c in n["cells"])
    return 0  # a preset node (should not appear after expand)


def depth(n: Node) -> int:
    if "element" in n:
        return 1
    kids = n["children"] if "children" in n else n["cells"] if "cells" in n else []
    return 1 + max((depth(k) for k in kids), default=0) if kids else 1
